core/services/keeper/registry_synchronizer_process_logs.go

```diff
@@ -0,0 +1,193 @@
+	defer done()
+	oldEnough := isOldEnoughConstructor(head, rs.minConfirmations)
+	i := rs.mailRoom.mbSyncRegistry.RetrieveIf(oldEnough)
+	if i == nil {
+		return
+	}
+	broadcast, ok := i.(log.Broadcast)
+	if !ok {
+		logger.Errorf("RegistrySynchronizer: invariant violation, expected log.Broadcast but got %T", broadcast)
+		return
+	}
+	was, err := broadcast.WasAlreadyConsumed()
+	if err != nil {
+		logger.Warn(err)
+		return
+	}
+	if was {
+		return
+	}
+	if !head.IsInChain(broadcast.RawLog().BlockHash) {
+		return
+	}
+	_, err = rs.syncRegistry()
+	if err != nil {
+		logger.Error(err)
+		return
+	}
+	err = broadcast.MarkConsumed()
+	logger.ErrorIf(err)
+}
+
+func (rs *RegistrySynchronizer) handleUpkeepCanceledLogs(head models.Head, done func()) {
+	defer done()
+	oldEnough := isOldEnoughConstructor(head, rs.minConfirmations)
+	for {
+		i := rs.mailRoom.mbUpkeepCanceled.RetrieveIf(oldEnough)
+		if i == nil {
+			return
+		}
+		broadcast, ok := i.(log.Broadcast)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected log.Broadcast but got %T", broadcast)
+			continue
+		}
+		was, err := broadcast.WasAlreadyConsumed()
+		if err != nil {
+			logger.Warn(err)
+			continue
+		}
+		if was {
+			continue
+		}
+		if !head.IsInChain(broadcast.RawLog().BlockHash) {
+			continue
+		}
+		log, ok := broadcast.DecodedLog().(*keeper_registry_wrapper.KeeperRegistryUpkeepCanceled)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected UpkeepCanceled log but got %T", log)
+			continue
+		}
+		ctx, cancel := postgres.DefaultQueryCtx()
+		defer cancel()
+		err = rs.orm.BatchDeleteUpkeepsForJob(ctx, rs.job.ID, []int64{log.Id.Int64()})
+		if err != nil {
+			logger.Error(err)
+			continue
+		}
+		err = broadcast.MarkConsumed()
+		logger.ErrorIf(err)
+	}
+}
+
+func (rs *RegistrySynchronizer) handleUpkeepRegisteredLogs(head models.Head, done func()) {
+	defer done()
+	oldEnough := isOldEnoughConstructor(head, rs.minConfirmations)
+	ctx, cancel := postgres.DefaultQueryCtx()
+	defer cancel()
+	registry, err := rs.orm.RegistryForJob(ctx, rs.job.ID)
+	if err != nil {
+		logger.Error(err)
+		return
+	}
+	for {
+		i := rs.mailRoom.mbUpkeepRegistered.RetrieveIf(oldEnough)
+		if i == nil {
+			return
+		}
+		broadcast, ok := i.(log.Broadcast)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected log.Broadcast but got %T", broadcast)
+			continue
+		}
+		was, err := broadcast.WasAlreadyConsumed()
+		if err != nil {
+			logger.Warn(err)
+			continue
+		}
+		if was {
+			continue
+		}
+		if !head.IsInChain(broadcast.RawLog().BlockHash) {
+			continue
+		}
+		log, ok := broadcast.DecodedLog().(*keeper_registry_wrapper.KeeperRegistryUpkeepRegistered)
+		if !ok {
+			logger.Errorf("RegistrySynchronizer: invariant violation, expected UpkeepRegistered log but got %T", log)
+			continue
+		}
+		err = rs.syncUpkeep(registry, log.Id.Int64())
+		if err != nil {
+			logger.Error(err)
+			continue
+		}
```

Comment:
though i suppose you could create a common method e.g. 
`func handleBroadcast(mailbox, handle func(broadcast))`
and call it with those lines wrapped in a func:

```
		log, ok := broadcast.DecodedLog().(*keeper_registry_wrapper.KeeperRegistryUpkeepRegistered)
		if !ok {
			logger.Errorf("RegistrySynchronizer: invariant violation, expected UpkeepRegistered log but got %T", log)
			continue
		}
		err = rs.syncUpkeep(registry, log.Id.Int64())
```

Reply:
I think you're right I could find a way to dry some of this up. Personally, I don't know that it's worth the complexity tradeoff, esp since we're going to remove this code entirely in the near future. I'd rather just keep it explicit for now.

Reply:
ok 👍🏼